// @mikro-orm/core/entity/EntityFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityFactory = void 0;
const Utils_1 = require("../utils/Utils");
const enums_1 = require("../enums");
class EntityFactory {
    constructor(unitOfWork, em) {
        this.unitOfWork = unitOfWork;
        this.em = em;
        this.driver = this.em.getDriver();
        this.config = this.em.config;
        this.metadata = this.em.getMetadata();
        this.hydrator = this.config.getHydrator(this, this.em);
    }
    create(entityName, data, options = {}) {
        var _a;
        options.initialized = (_a = options.initialized) !== null && _a !== void 0 ? _a : true;
        if (Utils_1.Utils.isEntity(data)) {
            return data;
        }
        entityName = Utils_1.Utils.className(entityName);
        const meta = this.metadata.get(entityName);
        meta.primaryKeys.forEach(pk => this.denormalizePrimaryKey(data, pk, meta.properties[pk]));
        const entity = this.createEntity(data, meta, options.merge, options.convertCustomTypes);
        const wrapped = entity.__helper;
        if (options.initialized) {
            this.hydrator.hydrate(entity, meta, data, options.newEntity, options.convertCustomTypes);
        }
        else {
            this.hydrator.hydrateReference(entity, meta, data, options.convertCustomTypes);
        }
        if (options.merge) {
            this.unitOfWork.merge(entity, undefined, false);
        }
        wrapped.__initialized = options.initialized;
        this.runHooks(entity, meta);
        return entity;
    }
    createReference(entityName, id, options = {}) {
        var _a;
        options.convertCustomTypes = (_a = options.convertCustomTypes) !== null && _a !== void 0 ? _a : true;
        entityName = Utils_1.Utils.className(entityName);
        const meta = this.metadata.get(entityName);
        if (Array.isArray(id)) {
            id = Utils_1.Utils.getPrimaryKeyCondFromArray(id, meta.primaryKeys);
        }
        const pks = Utils_1.Utils.getOrderedPrimaryKeys(id, meta, this.driver.getPlatform(), options.convertCustomTypes);
        if (Utils_1.Utils.isPrimaryKey(id)) {
            id = { [meta.primaryKeys[0]]: id };
        }
        const exists = this.unitOfWork.getById(entityName, pks);
        if (exists) {
            return exists;
        }
        return this.create(entityName, id, Object.assign({ initialized: false }, options));
    }
    createEntity(data, meta, merge, convertCustomTypes) {
        meta = this.processDiscriminatorColumn(meta, data);
        const Entity = meta.class;
        if (meta.primaryKeys.some(pk => !Utils_1.Utils.isDefined(data[pk], true))) {
            const params = this.extractConstructorParams(meta, data);
            meta.constructorParams.forEach(prop => delete data[prop]);
            // creates new instance via constructor as this is the new entity
            return new Entity(...params);
        }
        const pks = Utils_1.Utils.getOrderedPrimaryKeys(data, meta, this.driver.getPlatform(), convertCustomTypes);
        const exists = this.unitOfWork.getById(meta.name, pks);
        if (exists) {
            return exists;
        }
        // creates new entity instance, bypassing constructor call as its already persisted entity
        const entity = Object.create(Entity.prototype);
        entity.__helper.__managed = true;
        this.hydrator.hydrateReference(entity, meta, data, convertCustomTypes);
        if (merge) {
            this.unitOfWork.merge(entity, new Set([entity]), false);
        }
        return entity;
    }
    processDiscriminatorColumn(meta, data) {
        const root = Utils_1.Utils.getRootEntity(this.metadata, meta);
        if (!root.discriminatorColumn) {
            return meta;
        }
        const prop = meta.properties[root.discriminatorColumn];
        const value = data[prop.name];
        const type = root.discriminatorMap[value];
        meta = type ? this.metadata.find(type) : meta;
        // `prop.userDefined` is either `undefined` or `false`
        if (prop.userDefined === false) {
            delete data[prop.name];
        }
        return meta;
    }
    /**
     * denormalize PK to value required by driver (e.g. ObjectId)
     */
    denormalizePrimaryKey(data, primaryKey, prop) {
        const platform = this.driver.getPlatform();
        const pk = platform.getSerializedPrimaryKeyField(primaryKey);
        if (Utils_1.Utils.isDefined(data[pk], true) || Utils_1.Utils.isDefined(data[primaryKey], true)) {
            let id = data[pk] || data[primaryKey];
            if (prop.type.toLowerCase() === 'objectid') {
                id = platform.denormalizePrimaryKey(id);
            }
            delete data[pk];
            data[primaryKey] = id;
        }
    }
    /**
     * returns parameters for entity constructor, creating references from plain ids
     */
    extractConstructorParams(meta, data) {
        return meta.constructorParams.map(k => {
            if (meta.properties[k] && [enums_1.ReferenceType.MANY_TO_ONE, enums_1.ReferenceType.ONE_TO_ONE].includes(meta.properties[k].reference) && data[k]) {
                const entity = this.unitOfWork.getById(meta.properties[k].type, data[k]);
                if (entity) {
                    return entity;
                }
                if (Utils_1.Utils.isEntity(data[k])) {
                    return data[k];
                }
                return this.createReference(meta.properties[k].type, data[k]);
            }
            if (!meta.properties[k]) {
                return data;
            }
            return data[k];
        });
    }
    runHooks(entity, meta) {
        var _a;
        /* istanbul ignore next */
        const hooks = ((_a = meta.hooks) === null || _a === void 0 ? void 0 : _a.onInit) || [];
        if (hooks.length > 0) {
            hooks.forEach(hook => entity[hook]());
        }
        this.em.getEventManager().dispatchEvent(enums_1.EventType.onInit, { entity, em: this.em });
    }
}
exports.EntityFactory = EntityFactory;
