// @mikro-orm/core/entity/Reference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Reference = void 0;
const wrap_1 = require("./wrap");
class Reference {
    constructor(entity) {
        this.entity = entity;
        this.set(entity);
        const wrapped = this.entity.__helper;
        Object.defineProperty(this, '__reference', { value: true });
        wrapped.__meta.primaryKeys.forEach(primaryKey => {
            Object.defineProperty(this, primaryKey, {
                get() {
                    return this.entity[primaryKey];
                },
            });
        });
        if (wrapped.__meta.serializedPrimaryKey && wrapped.__meta.primaryKeys[0] !== wrapped.__meta.serializedPrimaryKey) {
            Object.defineProperty(this, wrapped.__meta.serializedPrimaryKey, {
                get() {
                    return this.entity.__helper.__serializedPrimaryKey;
                },
            });
        }
    }
    static create(entity) {
        if (Reference.isReference(entity)) {
            return entity;
        }
        return new Reference(entity);
    }
    /**
     * Checks whether the argument is instance or `Reference` wrapper.
     */
    static isReference(data) {
        return data && !!data.__reference;
    }
    /**
     * Wraps the entity in a `Reference` wrapper if the property is defined as `wrappedReference`.
     */
    static wrapReference(entity, prop) {
        if (entity && prop.wrappedReference && !Reference.isReference(entity)) {
            return Reference.create(entity);
        }
        return entity;
    }
    /**
     * Returns wrapped entity.
     */
    static unwrapReference(ref) {
        return Reference.isReference(ref) ? ref.unwrap() : ref;
    }
    async load(prop) {
        if (!this.isInitialized()) {
            await this.entity.__helper.init();
        }
        if (prop) {
            return this.entity[prop];
        }
        return this.entity;
    }
    set(entity) {
        if (entity instanceof Reference) {
            entity = entity.unwrap();
        }
        this.entity = entity;
        Object.defineProperty(this, '__helper', { value: this.entity.__helper, writable: true });
        Object.defineProperty(this, '$', { value: this.entity, writable: true });
        Object.defineProperty(this, 'get', { value: () => this.entity, writable: true });
    }
    unwrap() {
        return this.entity;
    }
    getEntity() {
        if (!this.isInitialized()) {
            throw new Error(`Reference<${this.entity.__helper.__meta.name}> ${this.entity.__helper.__primaryKey} not initialized`);
        }
        return this.entity;
    }
    getProperty(prop) {
        return this.getEntity()[prop];
    }
    isInitialized() {
        return this.entity.__helper.isInitialized();
    }
    populated(populated) {
        this.entity.__helper.populated(populated);
    }
    toJSON(...args) {
        return wrap_1.wrap(this.entity).toJSON(...args);
    }
}
exports.Reference = Reference;
